### PR TITLE
Word2 vecとbert new

### PR DESCRIPTION
**Pull request by maton369**

### CODE REVIEW

**Review by maton369 — COMMENTED** [review]
PRレビュー: 「リスト 5.3.6 Embedding Matrix作成」— 理論面を中心に

## サマリ
Keras の `Tokenizer` が作る語彙（1始まりのID割当て）と、事前学習 Word2Vec のベクトル空間を結合して埋め込み行列 `embedding_matrix` を構築する処理。全体の流れは妥当ですが、**gensim 4系の互換性**・**語彙整合性（OOV/未登録語）**・**次元整合性**・**メモリ計画**に関していくつかブロッカーと改善余地があります。

---

## Blocking（必ず直したい）
1) **gensim 4 互換性（`wv.vocab` の廃止）**  
　現行コードは `if word in word_vectors.wv.vocab:` を使用していますが、gensim 4 では `wv.vocab` が削除され、`key_to_index` に置き換わりました。KeyedVectors を使うなら
- 会員判定: `if word in kv.key_to_index:`  
- ベクトル取得: `kv.get_vector(word)`（`kv[word]` でも可）  
に全面置換が必要です。互換層を挟むと安全です（後述の提案コード参照）。

2) **次元の整合性チェックがない**  
　`EMBEDDING_DIM` と `word_vectors.vector_size` が一致しないと、静かに破綻します（ゼロ初期化された列が混在等）。`assert EMBEDDING_DIM == kv.vector_size` を最初に入れて、**空間次元の同一性**を保証してください。  
　理論的には、異なる次元の埋め込み空間は同一線形空間上に定義されておらず（基底次元数が異なる）、重み行列にコピーできても、モデルは期待する情報幾何（コサイン類似度や内積の分布）を再現できません。

3) **語彙整合性（Tokenizer vs Word2Vec）**  
　日本語はトークナイズの仕方（形態素辞書・分かち書き）がベクトル学習時と推論時で異なると **OOV が激増**します。OOV が多いと、`embedding_matrix` の大半がゼロ列になり、Embedding 層が情報を持てず、分類器が上位層で表層頻度に引っ張られる形で過学習しがちです。  
　→ 可能なら **同一の分かち書き**（辞書・前処理ルール）で学習済みベクトルを用意、または fastText のようなサブワードモデルに切替えるのが理論的に堅いです。

---

## Major（強く推奨）
4) **OOV トークンと PAD トークンの明示設計**  
　- `Tokenizer(oov_token="<UNK>")` を使い、`<UNK>` の index を取得して、列を平均ベクトルやランダム正規化ベクトルで初期化する方が学習が安定します（完全ゼロは勾配が流れにくい）。  
　- PAD（index 0）はゼロ埋めで OK（マスク処理を入れるなら `mask_zero=True` も検討）。  
　**理論背景**: 未知語ベクトルをゼロ列にすると、埋め込み→線形変換で常にバイアス項のみが寄与（表現力が下がる）。平均ベクトル等で初期化すれば、**語彙空間の中心**近傍から微調整可能。

5) **語彙上限の導入とメモリ見積り**  
　`num_words = len(word_index)` は全語彙を対象にします。語彙が大きいと、`(num_words+1) × EMBEDDING_DIM × 4 bytes` が線形に膨張。  
　例: 10万語 × 300次元 × float32 ≈ **114MB**。学習時に複製（GPU転送・勾配）も発生します。  
　→ `Tokenizer(num_words=MAX_NUM_WORDS)` と `min(i, MAX_NUM_WORDS)` のガードで上位頻度語に絞るのが実践的です。

6) **初期化後の扱い（固定 or 微調整）**  
　Embedding 層を `trainable=False` にすると、**転移学習としての安定性**は高い一方、下流タスク語彙への適応は弱くなります。小規模データなら凍結・中〜大規模なら微調整（小さめ学習率）が経験的に良いです。理論的には、事前学習空間（一般ドメイン）と下流分布（特定ドメイン）に **分布シフト** がある時、上流埋め込みの**線形サブスペース調整**が有効です。

---

## Minor（改善提案）
7) **被覆率レポート**  
　「`embedding_matrix` に写像できた語の割合」をログに出すと、データ前処理の有効性指標になります（例: Coverage 72% など）。  
8) **正規化の一貫性**  
　Word2Vec はしばしば L2 正規化されていません。`Embedding` 後段が内積ベースならスケール差が学習を不安定化することがあります。`normalize=True` でコピーする、あるいは読み込み時に L2 正規化してから貼る設計も検討余地あり（ただし元の空間幾何が変わる点は留意）。

---

## 安全・4系対応の提案コード（差分でなく置換用の完成版）
（`word_vectors` が `KeyedVectors` でも `Word2Vec` でも動くラッパを含む）

# 前提: gensim==4 系, Keras(TensorFlow) の Tokenizer/語彙が既に得られている
import numpy as np
from gensim.models import KeyedVectors, Word2Vec

def as_keyed_vectors(obj) -> KeyedVectors:
    """
    Word2Vec/KeyedVectors いずれが来ても KeyedVectors を返す小ラッパ。
    gensim 4 系では vocab -> key_to_index に置換されている点に注意。
    """
    if isinstance(obj, KeyedVectors):
        return obj
    if isinstance(obj, Word2Vec):
        return obj.wv
    # 文字列パスが来た場合の補助（任意）
    if isinstance(obj, str):
        try:
            return KeyedVectors.load(obj, mmap='r')
        except Exception:
            # word2vec 互換: .bin は binary=True, .vec は False
            binary = obj.lower().endswith(".bin")
            return KeyedVectors.load_word2vec_format(obj, binary=binary)
    raise TypeError("Unsupported embeddings object")

def build_embedding_matrix(word_index: dict,
                           embeddings,
                           embedding_dim: int,
                           max_num_words: int | None = None,
                           oov_index: int | None = None,
                           oov_init: str = "mean"  # "zero" or "random" も可
                           ) -> np.ndarray:
    """
    Keras Tokenizer の語彙と事前学習ベクトルから埋め込み行列を構築。
    - index=0 は PAD 用にゼロ。
    - max_num_words があれば上位頻度語のみ採用。
    - OOV インデックスを指定された場合、平均/乱数/ゼロで初期化。
    """
    kv = as_keyed_vectors(embeddings)
    assert embedding_dim == kv.vector_size, \
        f"次元不一致: EMBEDDING_DIM={embedding_dim}, model={kv.vector_size}"

    # 語彙上限の決定
    vocab_size = len(word_index) if max_num_words is None else min(len(word_index), max_num_words)
    mat = np.zeros((vocab_size + 1, embedding_dim), dtype=np.float32)

    # 事前に平均ベクトルを計算しておく（OOV 初期化に使う場合）
    mean_vec = kv.vectors.mean(axis=0) if oov_init == "mean" else None

    hit = 0
    for word, idx in word_index.items():
        if idx == 0:
            continue  # PAD（安全側）
        if idx > vocab_size:
            continue  # 上限外は捨てる
        if word in kv.key_to_index:
            mat[idx] = kv.get_vector(word)
            hit += 1
        # それ以外（OOV）はゼロのまま（後で oov_index に対してだけ処理）
        # ※ OOV を個別単語ごとに乱数で埋める運用もあるが、再現性と一貫性のため通常は固定

    # OOV インデックスを特別に扱う（Tokenizer(oov_token=...) を使った場合）
    if oov_index is not None and oov_index <= vocab_size:
        if oov_init == "mean" and mean_vec is not None:
            mat[oov_index] = mean_vec
        elif oov_init == "random":
            # Xavier風のスケールで初期化（単純例）
            scale = np.sqrt(1.0 / embedding_dim)
            mat[oov_index] = np.random.uniform(-scale, scale, size=(embedding_dim,)).astype(np.float32)
        else:
            mat[oov_index] = 0.0  # zero

    coverage = hit / vocab_size if vocab_size else 0.0
    print(f"[Embedding] vocab_size={vocab_size}, hit={hit}, coverage={coverage:.1%}")
    return mat

# --- 使い方例 ---
# 例: Tokenizer を oov_token 付きで初期化していると仮定
# from tensorflow.keras.preprocessing.text import Tokenizer
# tokenizer = Tokenizer(num_words=MAX_NUM_WORDS, oov_token="<UNK>")
# word_index = tokenizer.word_index
# oov_index = word_index.get("<UNK>")  # 無ければ None

# kv = KeyedVectors.load("ja.kv", mmap='r')  # または load_word2vec_format("ja.vec", binary=False)
# EMBEDDING_DIM = kv.vector_size
# embedding_matrix = build_embedding_matrix(
#     word_index=word_index,
#     embeddings=kv,
#     embedding_dim=EMBEDDING_DIM,
#     max_num_words=MAX_NUM_WORDS,   # 上限を設けないなら None
#     oov_index=oov_index,
#     oov_init="mean"
# )

# その後の Keras 層例:
# from tensorflow.keras.layers import Embedding
# emb_layer = Embedding(
#     input_dim=embedding_matrix.shape[0],
#     output_dim=embedding_matrix.shape[1],
#     weights=[embedding_matrix],
#     input_length=MAX_LEN,
#     trainable=False  # 分布シフトが強ければ True + 小LR 微調整も検討
# )

---

## テスト観点（理論的妥当性の確認）
- **被覆率**が 60% 未満なら、分かち書き不一致・表記揺れ（全角半角・記号・活用）・ドメイン差を疑う。正規化（NFKC、数字正規化、英字 lower など）を見直す。  
- **コサイン分布**の sanity check（数百語をサンプルし `cosine(u,v)` の分布を確認）。極端な尖りや崩壊があれば正規化/スケーリングを調整。  
- **学習曲線**で Embedding 凍結と微調整を比較。小データなら凍結が安定しやすい（バリアントによっては fine-tune で 1〜3pt 改善）。

---

## 結論
- まずは **gensim 4 互換（`key_to_index` / `get_vector`）と次元アサート**を入れて堅牢化。  
- OOV と PAD の設計を明示し、**被覆率を定量化**して前処理を点検。  
- 大語彙では `num_words` 上限を設けてメモリ・学習安定性を確保。  
この3点を押さえると、埋め込み転移の理論的前提（同一空間・十分な被覆・安定した最適化）が満たされ、下流モデルの汎化が改善します。
